File: algorithms/searching/binary-search/order-agnostic/OrderAgnosticBinarySearch.py
def order_agnostic_binary_search(arr, target):
    if not arr:
        return -1
    start, end = 0, len(arr) - 1
    ascending = arr[start] < arr[end]
    while start <= end:
        mid = start + (end - start) // 2
        if arr[mid] == target:
            return mid
        if ascending:
            if target < arr[mid]:
                end = mid - 1
            else:
                start = mid + 1
        else:
            if target > arr[mid]:
                end = mid - 1
            else:
                start = mid + 1
    return -1

File: algorithms/searching/binary-search/order-agnostic/OrderAgnosticBinarySearch.rs
fn orderAgnosticBinarySearch(arr: &[i32], target: i32) -> i32 {
    if arr.len() == 0 {
        return -1;
    }
    let mut start: i32 = 0;
    let mut end: i32 = arr.len() as i32 - 1;
    let isAscending: bool = arr[start as usize] < arr[end as usize];
    while start <= end {
        let mid: i32 = start + (end - start) / 2;
        if arr[mid as usize] == target {
            return mid;
        }
        if isAscending {
            if target < arr[mid as usize] {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        } else {
            if target > arr[mid as usize] {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
    }
    return -1;
}
